File: problem2-dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        self.xDepth = 0
        self.yDepth = 0
        self.xParent = None
        self.yParent = None
        
        self.dfs(root, None, 0, x , y)
        return self.xDepth == self.yDepth and self.xParent != self.yParent
    
    def dfs(self, root, parent, depth, x, y):
        #base
        if root == None: return 
        #logic
        if root.val == x:
            self.xDepth = depth
            self.xParent = parent
        if root.val == y:
            self.yDepth = depth 
            self.yParent = parent
        
        self.dfs(root.left, root, depth+1, x, y)
        self.dfs(root.right, root, depth+1, x, y)
            
#Time complexity O(n)
#Space complexity O(h)
#Using DFS to record child and parent value and comparing if the child values are at same level and parents are different or not.
        

File: Problem2_DFS.java
//Time Complexity: O(n)
//Space Complexity: O(h); Recursion Stack Space
//Code run successfully on LeetCode.

public class Problem2_DFS {

    int x_depth;
    int y_depth;
    TreeNode x_parent;
    TreeNode y_parent;
    
    public boolean isCousins(TreeNode root, int x, int y) {
        
        if(root == null)
            return false;
        
        dfs(root, 0 ,null,x,y);
        
        return ((x_depth == y_depth) && (x_parent != y_parent));
    }
    
    private void dfs(TreeNode root, int lvl, TreeNode parent, int x, int y){
        
        if(root == null)
            return;
        
        if(root.val == x){
            x_depth = lvl;
            x_parent = parent;
        }
        
        else if(root.val == y){
            y_depth = lvl;
            y_parent = parent;
        }
        
        if(x_parent != null && y_parent!= null)
            return;
        
        dfs(root.left, lvl +1, root, x, y);
        
         if(x_parent != null && y_parent!= null)
            return;
        
        dfs(root.right, lvl +1, root, x, y);
    }
}
